# Naïve Bayes

In [1]:
import pandas as pd
import numpy as np
import pandas as pd

## Exercício 1
###  First Question
Dada a tabela ao lado, se uma pessoa tem 35 anos e ganha $50.000,00 (renda media), ele compra um computador?

In [2]:
import pandas as pd
  
data = {'rec':           ['r1',          'r2',       'r3',     'r4',    'r5',         'r6',          'r7',      'r8',    'r9',   'r10',         'r11',       'r12',     'r13',        'r14'],
        'Age':           ['<=30',      '<=30',  '31...40',    '>40',   '>40',        '>40',     '31...40',    '<=30',  '<=30',   '>40',        '<=30',   '31...40', '31...40',        '>40'],
        'Income':        ['High',      'High',     'High', 'Medium',   'Low',        'Low',         'Low',  'Medium',   'Low','Medium',      'Medium',    'Medium',    'High',     'Medium'],
        'Student':       [  'No',        'No',       'No',     'No',   'Yes',        'Yes',         'Yes',      'No',   'Yes',   'Yes',         'Yes',        'No',     'Yes',         'No'],
        'Credit_rating': ['Fair', 'Excellent',     'Fair',   'Fair',  'Fair',  'Excellent',   'Excellent',    'Fair',  'Fair',  'Fair',   'Excellent', 'Excellent',    'Fair',  'Excellent'],
        'Buys_computer': [  'No',        'No',      'Yes',    'Yes',   'Yes',         'No',         'Yes',      'No',   'Yes',   'Yes',         'Yes',       'Yes',     'Yes',         'No'],
        }

df = pd.DataFrame (data, columns = ['rec','Age','Income','Student','Credit_rating','Buys_computer'])

df

rec      Age  Income Student Credit_rating Buys_computer
0    r1     <=30    High      No          Fair            No
1    r2     <=30    High      No     Excellent            No
2    r3  31...40    High      No          Fair           Yes
3    r4      >40  Medium      No          Fair           Yes
4    r5      >40     Low     Yes          Fair           Yes
5    r6      >40     Low     Yes     Excellent            No
6    r7  31...40     Low     Yes     Excellent           Yes
7    r8     <=30  Medium      No          Fair            No
8    r9     <=30     Low     Yes          Fair           Yes
9   r10      >40  Medium     Yes          Fair           Yes
10  r11     <=30  Medium     Yes     Excellent           Yes
11  r12  31...40  Medium      No     Excellent           Yes
12  r13  31...40    High     Yes          Fair           Yes
13  r14      >40  Medium      No     Excellent            No

**IMPORTANTE**: Este exercício será solucionado de duas maneiras:<br>

1º Solução) Considerando P=0 para os casos em que não existir nenhuma amostra que atenda à condição analisada;<br>

2º Solução) Utilizando a fórmula onde:<br>
P= (nc+mxp)/(n+m)<br>
Considerando que m=1 (peso dado à priori)<br>
nc - quantidade de amostras que atendem à condição<br>
p=m/(total de casos)=1/14<br>
n - depende da condição analisada, será igual ao número total de casos da condição <i>yes</i> OU
número total de casos da condição <i>no</i>.<br>


#### 1º Solução

In [3]:
#Definição das funções
def prob_yes(column,cond):
    """Probability of buy a computer with the input conditions"""
    if ((((df.loc[(df['Buys_computer']=='Yes') & (df[f'{column}']==f'{cond}'),f'{column}']).value_counts()).any())==False):
        x= 0
    else:
        x= df.loc[(df['Buys_computer']=='Yes') & (df[f'{column}']==f'{cond}'),f'{column}'].value_counts()[0]/df[['Buys_computer']].value_counts()[0]
    return(x)

def prob_no(column,cond):
    """Probability of not buy a computer with the input conditions"""
    if ((((df.loc[(df['Buys_computer']=='No') & (df[f'{column}']==f'{cond}'),f'{column}']).value_counts()).any())==False):
        x= 0
    else:
        x= df.loc[(df['Buys_computer']=='No') & (df[f'{column}']==f'{cond}'),f'{column}'].value_counts()[0]/df[['Buys_computer']].value_counts()[1]
    return(x)

#Cálculo das probabilidades de comprar ou não um computador
p_buy     = df[['Buys_computer']].value_counts()[0]/df['Buys_computer'].shape[0]
p_not_buy = df[['Buys_computer']].value_counts()[1]/df['Buys_computer'].shape[0]

In [4]:
#Solução:
print('x’ = (Age=31...40,Income=Medium)')
print('\n')

#P_buy
print('P(Age=31...40 | Buys_computer=Yes)=',  prob_yes('Age','31...40'))
print('P(Income=Medium | Buys_computer=Yes)=',prob_yes('Income','Medium'))
print('P(Buys_computer=Yes)=',p_buy)
print('\n')

#P_not_buy
print('P(Age=31...40 | Buys_computer=No)=',  prob_no('Age','31...40'))
print('P(Income=Medium | Buys_computer=No)=',prob_no('Income','Medium'))
print('P(Buys_computer=No)=',p_not_buy)
print('\n')

#Resultado:
p_yes = prob_yes('Age','31...40')*prob_yes('Income','Medium')*p_buy
p_no = prob_no('Age','31...40')*prob_no('Income','Medium')*p_not_buy
print('P(Yes|x’)=',p_yes)
print('P(No|x’)=',p_no)
if(p_yes>p_no):
    print("Yes, he buys a computer.")
else:
    print("No, he didn't buy a computer.")

x’ = (Age=31...40,Income=Medium)


P(Age=31...40 | Buys_computer=Yes)= 0.4444444444444444
P(Income=Medium | Buys_computer=Yes)= 0.4444444444444444
P(Buys_computer=Yes)= 0.6428571428571429


P(Age=31...40 | Buys_computer=No)= 0
P(Income=Medium | Buys_computer=No)= 0.4
P(Buys_computer=No)= 0.35714285714285715


P(Yes|x’)= 0.12698412698412698
P(No|x’)= 0.0
Yes, he buys a computer.


#### 2º Solução

In [5]:
#Definição das funções
def prob_yes2(column,cond):
    """Probability of buy a computer with the input conditions"""
    if ((((df.loc[(df['Buys_computer']=='Yes') & (df[f'{column}']==f'{cond}'),f'{column}']).value_counts()).any())==False):
        x= ((0 + 1/14)/(9+1))
    else:
        x= ((df.loc[(df['Buys_computer']=='Yes') & (df[f'{column}']==f'{cond}'),f'{column}'].value_counts()[0])+1/14)/(df[['Buys_computer']].value_counts()[0]+1)
    return(x)

def prob_no2(column,cond):
    """Probability of not buy a computer with the input conditions"""
    if ((((df.loc[(df['Buys_computer']=='No') & (df[f'{column}']==f'{cond}'),f'{column}']).value_counts()).any())==False):
        x= ((0 + 1/14)/(5+1))
    else:
        x= ((df.loc[(df['Buys_computer']=='No') & (df[f'{column}']==f'{cond}'),f'{column}'].value_counts()[0])+1/14)/(df[['Buys_computer']].value_counts()[1]+1)
    return(x)

#Cálculo das probabilidades de comprar ou não um computador
p_buy     = df[['Buys_computer']].value_counts()[0]/df['Buys_computer'].shape[0]
p_not_buy = df[['Buys_computer']].value_counts()[1]/df['Buys_computer'].shape[0]

In [6]:
#Solução:
print('x’ = (Age=31...40,Income=Medium)')
print('\n')

#P_buy
print('P(Age=31...40 | Buys_computer=Yes)=',  prob_yes2('Age','31...40'))
print('P(Income=Medium | Buys_computer=Yes)=',prob_yes2('Income','Medium'))
print('P(Buys_computer=Yes)=',p_buy)
print('\n')

#P_not_buy
print('P(Age=31...40 | Buys_computer=No)=',  prob_no2('Age','31...40'))
print('P(Income=Medium | Buys_computer=No)=',prob_no2('Income','Medium'))
print('P(Buys_computer=No)=',p_not_buy)
print('\n')

#Resultado:
p_yes = prob_yes2('Age','31...40')*prob_yes2('Income','Medium')*p_buy
p_no = prob_no2('Age','31...40')*prob_no2('Income','Medium')*p_not_buy
print('P(Yes|x’)=',p_yes)
print('P(No|x’)=',p_no)
if(p_yes>p_no):
    print("Yes, he buys a computer.")
else:
    print("No, he didn't buy a computer.")

x’ = (Age=31...40,Income=Medium)


P(Age=31...40 | Buys_computer=Yes)= 0.40714285714285714
P(Income=Medium | Buys_computer=Yes)= 0.40714285714285714
P(Buys_computer=Yes)= 0.6428571428571429


P(Age=31...40 | Buys_computer=No)= 0.011904761904761904
P(Income=Medium | Buys_computer=No)= 0.3452380952380953
P(Buys_computer=No)= 0.35714285714285715


P(Yes|x’)= 0.1065634110787172
P(No|x’)= 0.0014678490443796569
Yes, he buys a computer.


###  Second Question
E se a pessoa for estudante e tiver um crédito bom (fair)? 

#### 1º Solução

In [7]:
#Solução:
print('x’ = (Age=31...40, Income=Medium, Student=Yes, Credit_rating=Fair)')
print('\n')

#P_buy
print('P(Age=31...40 | Buys_computer=Yes)=',  prob_yes('Age','31...40'))
print('P(Income=Medium | Buys_computer=Yes)=',prob_yes('Income','Medium'))
print('P(Student=Yes | Buys_computer=Yes)=',prob_yes('Student','Yes'))
print('P(Credit_rating=Fair | Buys_computer=Yes)=',prob_yes('Credit_rating','Fair'))
print('P(Buys_computer=Yes)=',p_buy)
print('\n')

#P_not_buy
print('P(Age=31...40 | Buys_computer=No)=',  prob_no('Age','31...40'))
print('P(Income=Medium | Buys_computer=No)=',prob_no('Income','Medium'))
print('P(Student=Yes | Buys_computer=No)=',prob_no('Student','Yes'))
print('P(Credit_rating=Fair | Buys_computer=No)=',prob_no('Credit_rating','Fair'))
print('P(Buys_computer=No)=',p_not_buy)
print('\n')

#Resultado:
p_yes = prob_yes('Age','31...40')*prob_yes('Income','Medium')*prob_yes('Student','Yes')*prob_yes('Credit_rating','Fair')*p_buy
p_no = prob_no('Age','31...40')*prob_no('Income','Medium')*prob_no('Student','Yes')*prob_no('Credit_rating','Fair')*p_not_buy
print('P(Yes|x’)=',p_yes)
print('P(No|x’)=',p_no)
if(p_yes>p_no):
    print("Yes, he buys a computer.")
else:
    print("No, he didn't buy a computer.")

x’ = (Age=31...40, Income=Medium, Student=Yes, Credit_rating=Fair)


P(Age=31...40 | Buys_computer=Yes)= 0.4444444444444444
P(Income=Medium | Buys_computer=Yes)= 0.4444444444444444
P(Student=Yes | Buys_computer=Yes)= 0.6666666666666666
P(Credit_rating=Fair | Buys_computer=Yes)= 0.6666666666666666
P(Buys_computer=Yes)= 0.6428571428571429


P(Age=31...40 | Buys_computer=No)= 0
P(Income=Medium | Buys_computer=No)= 0.4
P(Student=Yes | Buys_computer=No)= 0.2
P(Credit_rating=Fair | Buys_computer=No)= 0.4
P(Buys_computer=No)= 0.35714285714285715


P(Yes|x’)= 0.056437389770723094
P(No|x’)= 0.0
Yes, he buys a computer.


#### 2º Solução

In [8]:
#Solução:
print('x’ = (Age=31...40, Income=Medium, Student=Yes, Credit_rating=Fair)')
print('\n')

#P_buy
print('P(Age=31...40 | Buys_computer=Yes)=',  prob_yes2('Age','31...40'))
print('P(Income=Medium | Buys_computer=Yes)=',prob_yes2('Income','Medium'))
print('P(Student=Yes | Buys_computer=Yes)=',prob_yes2('Student','Yes'))
print('P(Credit_rating=Fair | Buys_computer=Yes)=',prob_yes2('Credit_rating','Fair'))
print('P(Buys_computer=Yes)=',p_buy)
print('\n')

#P_not_buy
print('P(Age=31...40 | Buys_computer=No)=',  prob_no2('Age','31...40'))
print('P(Income=Medium | Buys_computer=No)=',prob_no2('Income','Medium'))
print('P(Student=Yes | Buys_computer=No)=',prob_no2('Student','Yes'))
print('P(Credit_rating=Fair | Buys_computer=No)=',prob_no2('Credit_rating','Fair'))
print('P(Buys_computer=No)=',p_not_buy)
print('\n')

#Resultado:
p_yes = prob_yes2('Age','31...40')*prob_yes2('Income','Medium')*prob_yes2('Student','Yes')*prob_yes2('Credit_rating','Fair')*p_buy
p_no = prob_no2('Age','31...40')*prob_no2('Income','Medium')*prob_no2('Student','Yes')*prob_no2('Credit_rating','Fair')*p_not_buy
print('P(Yes|x’)=',p_yes)
print('P(No|x’)=',p_no)
if(p_yes>p_no):
    print("Yes, he buys a computer.")
else:
    print("No, he didn't buy a computer.")

x’ = (Age=31...40, Income=Medium, Student=Yes, Credit_rating=Fair)


P(Age=31...40 | Buys_computer=Yes)= 0.40714285714285714
P(Income=Medium | Buys_computer=Yes)= 0.40714285714285714
P(Student=Yes | Buys_computer=Yes)= 0.6071428571428571
P(Credit_rating=Fair | Buys_computer=Yes)= 0.6071428571428571
P(Buys_computer=Yes)= 0.6428571428571429


P(Age=31...40 | Buys_computer=No)= 0.011904761904761904
P(Income=Medium | Buys_computer=No)= 0.3452380952380953
P(Student=Yes | Buys_computer=No)= 0.17857142857142858
P(Credit_rating=Fair | Buys_computer=No)= 0.3452380952380953
P(Buys_computer=No)= 0.35714285714285715


P(Yes|x’)= 0.03928166556345569
P(No|x’)= 9.049239431762341e-05
Yes, he buys a computer.


## Exercício 2

Use o Naïve Bayes Gaussiano para classificar o dataset Iris, e mais dois de sua escolha (os mesmos do trabalho anterior).

In [9]:
from sklearn.datasets import load_iris

### Definição da função principal utilizada

In [ ]:
#Função para calcular a probabilidade
def prob_calc(std,mean,feature):
    """Cálculo da probabilidade de valores contínuos.
    Input: std - standard deviation of feature and class;
           mean - mean of features and class;
           feature - values of input on feature column
    Output: Probability of all values"""
    z=[]
    for x in feature:
        z.append((1/(std*(2*np.pi)**(1/2))) * (np.exp(-((x-mean)**2) / (2*std**2))))
    return z

### Dataset Iris
Classificate into 'setosa', 'versicolor' or 'virginica'.

In [10]:
#Load the dataset
iris = load_iris()
X = iris['data']
iris['target_names']
y = iris['target']

#Separate iris samples - 0-setosa, 1-versicolor, 2-virginica
setosa = X[y==0]
versicolor = X[y==1]
virginica = X[y==2]

In [12]:
#Cálculo P(setosa), P(versicolor), P(virginica)
P_set = P_ver = P_virg = len(y[y==0])/len(y)

#Cálculo das probabilidades de cada feature da classe setosa
f1=prob_calc(np.std(setosa[:,0]),np.mean(setosa[:,0]),X[:,0])
f2=prob_calc(np.std(setosa[:,1]),np.mean(setosa[:,1]),X[:,1])
f3=prob_calc(np.std(setosa[:,2]),np.mean(setosa[:,2]),X[:,2])
f4=prob_calc(np.std(setosa[:,3]),np.mean(setosa[:,3]),X[:,3])

P_setosa=[]
for i in range(len(f1)):
    P_setosa.append(f1[i]*f2[i]*f3[i]*f4[i]*P_set)
#P_setosa


#Cálculo das probabilidades de cada feature da classe versicolor
f1=prob_calc(np.std(versicolor[:,0]),np.mean(versicolor[:,0]),X[:,0])
f2=prob_calc(np.std(versicolor[:,1]),np.mean(versicolor[:,1]),X[:,1])
f3=prob_calc(np.std(versicolor[:,2]),np.mean(versicolor[:,2]),X[:,2])
f4=prob_calc(np.std(versicolor[:,3]),np.mean(versicolor[:,3]),X[:,3])

P_versicolor=[]
for i in range(len(f1)):
    P_versicolor.append(f1[i]*f2[i]*f3[i]*f4[i]*P_ver)
#P_versicolor


#Cálculo das probabilidades de cada feature da classe virginica
f1=prob_calc(np.std(virginica[:,2]),np.mean(virginica[:,0]),X[:,0])
f2=prob_calc(np.std(virginica[:,1]),np.mean(virginica[:,1]),X[:,1])
f3=prob_calc(np.std(virginica[:,2]),np.mean(virginica[:,2]),X[:,2])
f4=prob_calc(np.std(virginica[:,3]),np.mean(virginica[:,3]),X[:,3])

P_virginica=[]
for i in range(len(f1)):
    P_virginica.append(f1[i]*f2[i]*f3[i]*f4[i]*P_virg)
#P_virginica


#Solução - Comparação entre as probabilidades para fazer a predição
prob_0=P_setosa
prob_1=P_versicolor
prob_2=P_virginica

classification=[]
for i in range(len(prob_0)-1):
    if ((prob_0[i]>prob_1[i])&(prob_0[i]>prob_2[i])):
        classification.append(0)
    if ((prob_1[i]>prob_0[i])&(prob_1[i]>prob_2[i])):
        classification.append(1)
    if ((prob_2[i]>prob_0[i])&(prob_2[i]>prob_1[i])):
        classification.append(2)
print('Predição de y:\n',classification)
print('\n')
x=[]
c=0
e=0
for i in range(len(classification)):
    if (classification[i]==y[i]):
        c=c+1
    else:
        e=e+1
print('Porcentagem de acerto:',(c/y.shape[0])*100)
print('Porcentagem de erro:',(e/y.shape[0])*100)

Predição de y:
 [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]


Porcentagem de acerto: 95.33333333333334
Porcentagem de erro: 4.0


Interpretação do resultado: Podemos observar que a porcentagem de erro é muito pequena (4%) comparada à porcentagem de acerto(95.33%), e portanto, o método Naïve Bayes teve apresentou um resultado satisfatório para a classificação do dataset Iris.

### Dataset Wholesale customers data.csv
Classificate into regions (Regions - Lisnon, Oporto or Other (Nominal)).
Neste exercício, ao invés de comparar a classificação com a clusterização realizada no Exercício do K-Means, optei por classificar o DataSet em Regiões.

In [13]:
# Importando o dataset
X = pd.read_csv('data/Wholesale customers data.csv')

#classify into region:
region = X['Region']
X.drop(columns=['Region','Channel'], inplace=True)

#Separate classes into 03 samples
r1 = X[region==1]
r2 = X[region==2]
r3 = X[region==3]

In [15]:
#Cálculo de P(r1), P(r2), P(r3)
P_r1 = len(region[region==1])/len(region)
P_r2 = len(region[region==2])/len(region) 
P_r3 = len(region[region==3])/len(region) 

#Cálculo das probabilidades de cada feature da classe r1
f1=prob_calc(np.std(r1.iloc[:,0]),np.mean(r1.iloc[:,0]),X.iloc[:,0])
f2=prob_calc(np.std(r1.iloc[:,1]),np.mean(r1.iloc[:,1]),X.iloc[:,1])
f3=prob_calc(np.std(r1.iloc[:,2]),np.mean(r1.iloc[:,2]),X.iloc[:,2])
f4=prob_calc(np.std(r1.iloc[:,3]),np.mean(r1.iloc[:,3]),X.iloc[:,3])
f5=prob_calc(np.std(r1.iloc[:,4]),np.mean(r1.iloc[:,4]),X.iloc[:,4])
f6=prob_calc(np.std(r1.iloc[:,5]),np.mean(r1.iloc[:,5]),X.iloc[:,5])

Prob_r1=[]
for i in range(len(f1)):
    Prob_r1.append(f1[i]*f2[i]*f3[i]*f4[i]*f5[i]*f6[i]*P_r1)
#Prob_r1

#Cálculo das probabilidades de cada feature da classe r2
f1=prob_calc(np.std(r2.iloc[:,0]),np.mean(r2.iloc[:,0]),X.iloc[:,0])
f2=prob_calc(np.std(r2.iloc[:,1]),np.mean(r2.iloc[:,1]),X.iloc[:,1])
f3=prob_calc(np.std(r2.iloc[:,2]),np.mean(r2.iloc[:,2]),X.iloc[:,2])
f4=prob_calc(np.std(r2.iloc[:,3]),np.mean(r2.iloc[:,3]),X.iloc[:,3])
f5=prob_calc(np.std(r2.iloc[:,4]),np.mean(r2.iloc[:,4]),X.iloc[:,4])
f6=prob_calc(np.std(r2.iloc[:,5]),np.mean(r2.iloc[:,5]),X.iloc[:,5])

Prob_r2=[]
for i in range(len(f1)):
    Prob_r2.append(f1[i]*f2[i]*f3[i]*f4[i]*f5[i]*f6[i]*P_r2)
#Prob_r2


#Cálculo das probabilidades de cada feature da classe r3
f1=prob_calc(np.std(r3.iloc[:,2]),np.mean(r3.iloc[:,0]),X.iloc[:,0])
f2=prob_calc(np.std(r3.iloc[:,1]),np.mean(r3.iloc[:,1]),X.iloc[:,1])
f3=prob_calc(np.std(r3.iloc[:,2]),np.mean(r3.iloc[:,2]),X.iloc[:,2])
f4=prob_calc(np.std(r3.iloc[:,3]),np.mean(r3.iloc[:,3]),X.iloc[:,3])
f5=prob_calc(np.std(r3.iloc[:,4]),np.mean(r3.iloc[:,4]),X.iloc[:,4])
f6=prob_calc(np.std(r3.iloc[:,5]),np.mean(r3.iloc[:,5]),X.iloc[:,5])

Prob_r3=[]
for i in range(len(f1)):
    Prob_r3.append(f1[i]*f2[i]*f3[i]*f4[i]*f5[i]*f6[i]*P_r3)
#Prob_r3

#Solução - Comparação entre as probabilidades para fazer a predição
prob_1 = Prob_r1
prob_2 = Prob_r2
prob_3 = Prob_r3

classification=[]
for i in range(len(prob_1)):
    if ((prob_1[i]>prob_2[i])&(prob_1[i]>prob_3[i])):
        classification.append(1)
    if ((prob_2[i]>prob_1[i])&(prob_2[i]>prob_3[i])):
        classification.append(2)
    if ((prob_3[i]>prob_1[i])&(prob_3[i]>prob_2[i])):
        classification.append(3)
print(classification)
print('\n')

#Análise do resultado obtido
x=[]
c=0
e=0
for i in range(len(classification)):
    if (classification[i]==region[i]):
        c=c+1
    else:
        e=e+1
print('Porcentagem de acerto:',(c/region.shape[0])*100)
print('Porcentagem de erro:',(e/region.shape[0])*100)

[3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 1, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 1, 3, 3, 3, 3, 1, 3, 3, 1, 3, 1, 3, 1, 1, 1, 3, 3, 3, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 1, 1, 3, 3, 3, 1, 3, 3, 1, 3, 1, 2, 3, 3, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 1, 3, 3, 3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 2, 3, 1, 1, 3, 3, 3, 3, 3, 1, 3, 3, 3, 1, 3, 1, 3, 1, 3, 3, 1, 1, 3, 3, 1, 3, 1, 1, 1, 3, 3, 1, 1, 3, 3, 3, 1, 3, 3, 3, 3, 3, 3, 1, 3, 3, 1, 3, 3, 1, 3, 3, 3, 3, 3, 3, 1, 3, 1, 3, 1, 3, 3, 3, 3, 1, 3, 1, 3, 3, 3, 3, 3, 3, 1, 3, 1, 3, 3, 1, 3, 1, 1, 1, 3, 3, 1, 3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 3, 3, 3, 3, 1, 3, 2, 1, 1, 3, 3, 2, 3, 3, 1, 3, 3, 1, 1, 3, 3, 3, 3, 1, 1, 3, 2, 1, 3, 3, 3, 3, 3, 1, 3, 3, 1, 1, 3, 3, 3, 3, 1, 1, 3, 1, 3, 1, 3, 1, 3, 1, 3, 3, 1, 3, 3, 3, 3, 3, 1, 3, 3, 1, 3, 1, 3, 1, 1, 1, 3, 3, 3, 1, 3, 3, 3, 3, 3, 3, 3, 1, 1, 1, 1, 1, 3, 2, 3, 3, 1, 3, 1, 3, 1, 1, 3, 3, 3, 1, 1, 1, 1, 3, 3, 3, 1, 1, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 1, 2, 3, 1, 3, 1, 1, 3, 3, 1, 1, 3, 3, 1, 2, 1, 1, 3, 1, 1, 3, 3, 

Interpretação do resultado: Podemos observar que a porcentagem de erro é muito próxima à % de acerto. Portanto, neste caso não é recomendado classificar o dataset através de **regiões** utilizando o método Naive Bayes, uma vez que o resultado obtido não terá uma precisão satisfatória.

### Wine dataset 
Classificate into 'class_0', 'class_1' or 'class_2'.

In [16]:
#Import dataset
from sklearn.datasets import load_wine
wine = load_wine()
X = wine['data']
y = wine['target']

#Separate classes into 03 samples
c0 = X[y==0]
c1 = X[y==1]
c2 = X[y==2]

In [17]:
#Cálculo das probabilidades de saída de cada classe
P0 = len(c0)/len(y)
P1 = len(c1)/len(y)
P2 = len(c2)/len(y)

#Cálculo das probabilidades para a classe 0
p0 = prob_calc(np.std(c0),np.mean(c0),X[:,0])
p1 = prob_calc(np.std(c0),np.mean(c0),X[:,1])
p2 = prob_calc(np.std(c0),np.mean(c0),X[:,2])
p3 = prob_calc(np.std(c0),np.mean(c0),X[:,3])
p4 = prob_calc(np.std(c0),np.mean(c0),X[:,4])
p5 = prob_calc(np.std(c0),np.mean(c0),X[:,5])
p6 = prob_calc(np.std(c0),np.mean(c0),X[:,6])
p7 = prob_calc(np.std(c0),np.mean(c0),X[:,7])
p8 = prob_calc(np.std(c0),np.mean(c0),X[:,8])
p9 = prob_calc(np.std(c0),np.mean(c0),X[:,9])
p10= prob_calc(np.std(c0),np.mean(c0),X[:,10])
p11= prob_calc(np.std(c0),np.mean(c0),X[:,11])
p12= prob_calc(np.std(c0),np.mean(c0),X[:,12])
PC0=[]

for i in range(len(p0)):
    PC0.append(p0[i]*p1[i]*p2[i]*p3[i]*p4[i]*p5[i]*p6[i]*p7[i]*p8[i]*p9[i]*p10[i]*p11[i]*p12[i]*P0)
PC0

#Cálculo das probabilidades para a classe 1
p0 = prob_calc(np.std(c1),np.mean(c1),X[:,0])
p1 = prob_calc(np.std(c1),np.mean(c1),X[:,1])
p2 = prob_calc(np.std(c1),np.mean(c1),X[:,2])
p3 = prob_calc(np.std(c1),np.mean(c1),X[:,3])
p4 = prob_calc(np.std(c1),np.mean(c1),X[:,4])
p5 = prob_calc(np.std(c1),np.mean(c1),X[:,5])
p6 = prob_calc(np.std(c1),np.mean(c1),X[:,6])
p7 = prob_calc(np.std(c1),np.mean(c1),X[:,7])
p8 = prob_calc(np.std(c1),np.mean(c1),X[:,8])
p9 = prob_calc(np.std(c1),np.mean(c1),X[:,9])
p10= prob_calc(np.std(c1),np.mean(c1),X[:,10])
p11= prob_calc(np.std(c1),np.mean(c1),X[:,11])
p12= prob_calc(np.std(c1),np.mean(c1),X[:,12])
PC1=[]

for i in range(len(p0)):
    PC1.append(p0[i]*p1[i]*p2[i]*p3[i]*p4[i]*p5[i]*p6[i]*p7[i]*p8[i]*p9[i]*p10[i]*p11[i]*p12[i]*P1)
PC1


#Cálculo das probabilidades para a classe 2
p0 = prob_calc(np.std(c2),np.mean(c2),X[:,0])
p1 = prob_calc(np.std(c2),np.mean(c2),X[:,1])
p2 = prob_calc(np.std(c2),np.mean(c2),X[:,2])
p3 = prob_calc(np.std(c2),np.mean(c2),X[:,3])
p4 = prob_calc(np.std(c2),np.mean(c2),X[:,4])
p5 = prob_calc(np.std(c2),np.mean(c2),X[:,5])
p6 = prob_calc(np.std(c2),np.mean(c2),X[:,6])
p7 = prob_calc(np.std(c2),np.mean(c2),X[:,7])
p8 = prob_calc(np.std(c2),np.mean(c2),X[:,8])
p9 = prob_calc(np.std(c2),np.mean(c2),X[:,9])
p10= prob_calc(np.std(c2),np.mean(c2),X[:,10])
p11= prob_calc(np.std(c2),np.mean(c2),X[:,11])
p12= prob_calc(np.std(c2),np.mean(c2),X[:,12])
PC2=[]

for i in range(len(p0)):
    PC2.append(p0[i]*p1[i]*p2[i]*p3[i]*p4[i]*p5[i]*p6[i]*p7[i]*p8[i]*p9[i]*p10[i]*p11[i]*p12[i]*P2)
PC2

#Predição
prob_0 = PC0
prob_1 = PC1
prob_2 = PC2
classification=[]
for i in range(len(prob_0)):
    if ((prob_0[i]>prob_1[i])&(prob_0[i]>prob_2[i])):
        classification.append(0)
    if ((prob_1[i]>prob_0[i])&(prob_1[i]>prob_2[i])):
        classification.append(1)
    if ((prob_2[i]>prob_0[i])&(prob_2[i]>prob_1[i])):
        classification.append(2)
print('Predição de y:\n',classification)

print('\n')

#Análise do resultado obtido
x=[]
c=0
e=0
for i in range(len(classification)):
    if (classification[i]==y[i]):
        c=c+1
    else:
        e=e+1
print('Porcentagem de acerto:',(c/y.shape[0])*100)
print('Porcentagem de erro:',(e/y.shape[0])*100)

Predição de y:
 [0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 1, 1, 1, 1, 2, 1, 1, 2, 2, 0, 1, 1, 0, 0, 1, 1, 1, 2, 1, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 0, 1, 1, 2, 1, 2, 1, 1, 1, 2, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 1, 1, 1, 1, 2, 1, 1, 0, 0, 1, 1, 2, 1, 1, 1, 1, 2, 2, 2, 1, 0, 2, 1, 1, 2, 1, 2, 1, 1, 2, 2, 2, 1, 1, 1, 2, 2, 2, 0, 0, 1]


Porcentagem de acerto: 70.78651685393258
Porcentagem de erro: 29.213483146067414


Interpretação do resultado: Podemos observar que a porcentagem de erro é pequena comparada à porcentagem de acerto, e portanto, o método Naïve Bayes teve apresentou um resultado satisfatório para a classificação do dataset Wine.